E2_ESRGAN/lib/phase_1.py

```diff
@@ -0,0 +1,68 @@
+from functools import partial
```

Comment:
Is this code structure of having a different file per-training phase followed elsewhere? 
Instead you can split your original model file into one module each for G, D & the overall model. And add each training phase as a helper on the main model class - this way its easier to track the behavior of the model across its entire lifetime in the pipeline.

Reply:
Okay!

Reply:
I think, this should be kept for later, once all the basic PRs gets merged. Else, it will be difficult to restructure the code, since there some unmerged branches, which are dependent on each other. Once all of them gets merged, we can take a look into this. What do you say?
I've opened an issue to keep track of this. #30

Reply:
TBH, "it will cause merge conflicts" is never a good reason to keep the code in a bad state. In fact, its a non-trivial chunk of the effort we put into getting in code at Google :-).

The bigger reason why I want a refactor, is because the `Model` class in this file actually trains the `RRDBNet` object from `model.py`. In essence, we now have two classes that encapsulate the same essential object but do two actions on them (which kind-of feels against the spirit of OOP).

As a middle ground, might I suggest:
* Refactor the class in this file to just a `warmup_generator(...)` function that takes in a `RRDBNet` object along with settings etc, and trains it as required.
* Add a `TODO(captain-pool)` block to eventually merge this into the main class.

This is mainly also because all of the the members in the `Model` class in this file are only utilized & accessed in the `train` function, and never elsewhere (or ever again)- so there is not real need for a class.

Reply:
SG